src/asset_manager/asset.rs

```diff
@@ -0,0 +1,184 @@
+
+/// The asset type.
+/// Every asset (Mesh, Texture, ...) has to implement
+/// this type in order to be compatible with the `AssetLoader`.
+///
+/// ### The concept
+///
+/// An asset is loaded in two steps:
+///
+/// 1) Load asset data (`AssetData`)
+/// 2) Turn asset data into asset (`from_data`)
+///
+/// # Examples
+///
+/// Here is an examle implementation of an
+/// asset:
+///
+/// ```
+/// use std::io::{Read, BufRead, BufReader};
+///
+/// use amethyst::asset_manager::{Asset, ImportError};
+/// use amethyst::engine::Context;
+///
+/// struct Table {
+///     num_rows: u32,
+///     num_columns: u32,
+///     data: Box<[i32]>,
+/// }
+///
+/// struct WhitespaceTable;
+///
+/// impl Asset for Table {
+///     type Data = Self;
+///     type Error = (); // Should be some InconsistentSizeError (if data.len() != num_rows * num_columns)
+///
+///     fn from_data(data: Self, _: &mut Context) -> Result<Self, ()> {
+///         Ok(data) // Ommitted here: Check for size
+///     }
+/// }
+///
+/// impl AssetFormat for WhitespaceTable {
+///     fn file_extension() -> &'static str {
+///         "wst"
+///     }
+/// }
+///
+/// impl Import<WhitspaceTable::Data> for WhitespaceTable {
+///     fn import<R: Read>(stream: R) -> Result<WhitespaceTable, ImportError> {
+///         let mut num_columns = 0;
+///         let mut old_columns = 0;
+///         let mut values = Vec::new();
+///         let mut reader = BufReader::new(stream);
+///
+///         for line in reader.lines() {
+///             for word in line.split_whitespace() {
+///                 num_columns += 1;
+///                 values.push(word.parse().map_err(|x| ImportError::FormatError("Invalid integer".to_string()))?)
+///             }
+///             if (old_columns != 0 && num_columns != old_columns) {
+///                 return Err(ImportError::FormatError(format!("Number of columns not matching (previous line: {}, this line: {})",
+///                                                             old_columns, num_columns)));
+///             }
+///             old_columns = num_columns;
+///             num_columns = 0;
+///         }
+///
+///         if values.len() == 0 {
+///             return Err(ImportError::FormatError("Empty file".to_string()));
+///         }
+///
+///         Ok(WhitespaceTable {
+///             num_columns: old_columns,
+///             num_rows: values.len() / old_columns,
+///             data: values.into_boxed_slice(),
+///         })
+///     }
+/// }
+/// ```
+pub trait Asset: Sized {
+    /// The data type, an intermediate format.
+    /// This may also be `Self` if this asset does
+    /// not depend on `Context`.
+    type Data;
+    /// The error type that may be returned if
+    /// `from_data` fails.
+    type Error: Debug;
+
+    /// Create the asset from the data and the context (used to create buffers for the gpu).
+    fn from_data(data: Self::Data, context: &mut Context) -> Result<Self, Self::Error>;
```

Comment:
I understand your point concerning passing `context` directly, which simplifies handling (e.g for the AssetFuture). Personally, I would consider making `Context` an associated type also an option to consider. This avoids blocking the context and allows the user to pass own related data for loading, which might not be contained in the context provided by the engine. But that's not a blocker from my side, other opinions on regarding this topic would be appreciated!